tools/pipelines-tasks/AVDAppAttachPublish/main.ts

```diff
@@ -0,0 +1,98 @@
+import { ToolRunner } from 'azure-pipelines-task-lib/toolrunner';
+import tl = require('azure-pipelines-task-lib/task');
+import path = require('path');
+import armStorage = require('azure-pipelines-tasks-azure-arm-rest/azure-arm-storage');
+import { AzureRMEndpoint } from 'azure-pipelines-tasks-azure-arm-rest/azure-arm-endpoint';
+import { AzureEndpoint, StorageAccount } from 'azure-pipelines-tasks-azure-arm-rest/azureModels';
+import * as os from 'os';
+
+const HELPER_SCRIPT = path.join(__dirname, 'publishAVD.ps1');
+const TARGET_DLL = path.join(__dirname, '/AppAttachFrameworkDLL/AppAttachKernel.dll');
+
+function isNonEmpty(str: string): boolean {
+    return (!!str && !!str.trim());
+}
+
+function getResourceGroupNameFromUri(resourceUri: string): string {
+    if (isNonEmpty(resourceUri)) {
+        resourceUri = resourceUri.toLowerCase();
+        return resourceUri.substring(resourceUri.indexOf("resourcegroups/") + "resourcegroups/".length, resourceUri.indexOf("/providers"));
+    }
+
+    return "";
+}
+
+const getPowershellRunner = (scriptPath: string): ToolRunner => {
+	const powershellRunner: ToolRunner = tl.tool('powershell')
+		.arg('-NoLogo')
+		.arg('-NoProfile')
+		.arg('-NonInteractive')
+		.arg(['-ExecutionPolicy', 'Unrestricted']);
+
+	// Quote the script path to allow for spaces.
+	// Existing quotes need to be escaped.
+	powershellRunner.arg(`& '${scriptPath.replace("'", "''")}'`);
+	return powershellRunner;
+}
+
+async function run(): Promise<void> {
+    try {
+		let connectedServiceName = tl.getInput('ConnectedServiceNameARM', true)!;
+
+		let subscriptionID: string = tl.getEndpointDataParameter(connectedServiceName, 'subscriptionid', true)!;
+
+		let storageAccountName = tl.getInput('StorageAccount', true)!;
+		var azureEndpoint: AzureEndpoint = await new AzureRMEndpoint(connectedServiceName).getEndpoint();
+		let accessToken: string = (await azureEndpoint.applicationTokenCredentials.getToken()).toString();
+
+		const storageArmClient = new armStorage.StorageManagementClient(azureEndpoint.applicationTokenCredentials, (azureEndpoint.subscriptionID)!);
+		let storageAccount: StorageAccount = await storageArmClient.storageAccounts.get(storageAccountName);
+		let storageAccountResourceGroupName = getResourceGroupNameFromUri(storageAccount.id);
+		let accessKeys = await storageArmClient.storageAccounts.listKeys(storageAccountResourceGroupName, storageAccountName, null);
+		let accessKey: string = accessKeys[0];
+		let StorageAccountConnectionString: string = "DefaultEndpointsProtocol=https;AccountName="+storageAccountName+";AccountKey="+accessKey
+														+";EndpointSuffix=core.windows.net";
+		let username: string = (process.env.BUILD_REQUESTEDFOREMAIL)!;
```

Comment:
What is this username/email used for?

Reply:
To authenticate target AVD user while app attach through app attach framework

Reply:
I don't think you can assume that the person starting the pipeline also has access to the Azure resources, so it may cause issues if you use this to authenticate.

I can easily imagine someone with access to the Azure resources setting up the pipeline and allowing others to run it when needed, while not giving them direct access to the Azure resources.

Reply:
Agreed, the current system is: "Azure Subscription" field would show up the available subscription under loggedin user and establishing service connection to one of these subscription is made mandatory and this is an assumption that the logged in user has azure resources. A better fallback for service connection to non-logged in user is not being supported in this version but surely in the near plan.